File: Tests/RedlineCoreTests/CadenceTests.swift
// Cues about how the work is spread out. The rules worth pinning are the ones that keep it
// from becoming a nag: once per stretch, once per night, and nothing at all when the setting
// is off or when the data is old enough that the claim would be about the past.
import XCTest
@testable import RedlineCore

final class CadenceTests: XCTestCase {
    private var calendar: Calendar = {
        var c = Calendar(identifier: .gregorian)
        c.timeZone = TimeZone(identifier: "UTC")!
        return c
    }()

    private func entry(_ ts: Date) -> Entry {
        Entry(provider: "Claude", key: UUID().uuidString, ts: ts, model: "claude-sonnet-5",
              input: 100, output: 10, cacheRead: 0, cache5m: 0, cache1h: 0)
    }

    /// 2026-08-18 09:00:00 UTC
    private let base = Date(timeIntervalSince1970: 1_787_043_600)

    private func run(minutes: [Double]) -> [Entry] {
        minutes.map { entry(base.addingTimeInterval($0 * 60)) }
    }

    // MARK: - Shapes

    func testAGapLongerThanTheBreakSplitsTheStretch() {
        let entries = run(minutes: [0, 5, 10, 40, 45])
        let stretches = Cadence.stretches(entries)
        XCTAssertEqual(stretches.count, 2)
        XCTAssertEqual(stretches[0].length, 600)
        XCTAssertEqual(stretches[1].length, 300)
    }

    func testAShortPauseDoesNotSplitAStretch() {
        // A slow turn or a build is not a break.
        let entries = run(minutes: [0, 12, 24])
        XCTAssertEqual(Cadence.stretches(entries).count, 1)
    }

    func testStretchLengthIsMeasuredToTheLastActivityNotToNow() {
        let entries = run(minutes: [0, 10, 20, 30])
        let stretch = Cadence.stretches(entries).last
        XCTAssertEqual(stretch?.length, 1800,
                       "a counter that climbs while nothing happens is not a measurement")
    }

    func testCurrentStretchClosesOnceTheGapHasPassed() {
        let entries = run(minutes: [0, 10, 20, 30])
        XCTAssertNotNil(Cadence.current(entries, now: base.addingTimeInterval(35 * 60)))
        XCTAssertNil(Cadence.current(entries, now: base.addingTimeInterval(60 * 60)))
    }

    func testStreakCountsBackFromTodayAndStopsAtTheFirstGap() {
        var entries: [Entry] = []
        for day in [0, 1, 2, 4, 5] {
            entries.append(entry(base.addingTimeInterval(-Double(day) * 86400)))
        }
        XCTAssertEqual(Cadence.streak(entries, endingOn: base, calendar: calendar), 3)
    }

    func testStreakIsZeroWhenTodayHasNothing() {
        let entries = [entry(base.addingTimeInterval(-2 * 86400))]
        XCTAssertEqual(Cadence.streak(entries, endingOn: base, calendar: calendar), 0)
    }

    func testHourHistogramBucketsByLocalHour() {
        let entries = run(minutes: [0, 1, 120])
        let hours = Cadence.byHourOfDay(entries, calendar: calendar)
        XCTAssertEqual(hours[9], 220)
        XCTAssertEqual(hours[11], 110)
        XCTAssertEqual(hours.reduce(0, +), 330)
    }

    func testNightRollsAtFiveNotAtMidnight() {
        // 01:30 belongs to the evening that ran into it.
        let lateEvening = base.addingTimeInterval(14 * 3600)      // 23:00
        let afterMidnight = base.addingTimeInterval(16.5 * 3600)  // 01:30 next day
        XCTAssertEqual(Cadence.nightKey(for: lateEvening, calendar: calendar),
                       Cadence.nightKey(for: afterMidnight, calendar: calendar))
    }

    // MARK: - Cues

    private func config() -> Config {
        var c = Config()
        c.mindfulCues = true
        c.stretchMinutes = 90
        c.lateHour = 23
        c.streakDays = 7
        return c
    }

    func testCuesAreSilentWhenTheSettingIsOff() {
        var state = CadenceState()
        var off = config()
        off.mindfulCues = false
        let entries = run(minutes: Array(stride(from: 0, through: 240, by: 10)).map(Double.init))
        let cues = CadenceRules.evaluate(entries: entries, config: off,
                                         now: base.addingTimeInterval(240 * 60),
                                         calendar: calendar, state: &state)
        XCTAssertTrue(cues.isEmpty)
    }

    func testAStretchIsAnnouncedOnceAndThenAtTheNextMultiple() {
        var state = CadenceState()
        let minutes = Array(stride(from: 0.0, through: 200.0, by: 10.0))
        // At 100 minutes: one threshold reached
        let atFirst = CadenceRules.evaluate(
            entries: run(minutes: minutes.filter { $0 <= 100 }), config: config(),
            now: base.addingTimeInterval(105 * 60), calendar: calendar, state: &state)
        XCTAssertEqual(atFirst.count, 1)
        if case .stretch = atFirst[0].kind {} else { XCTFail("expected a stretch cue") }

        // Ten minutes later, still the same run: nothing new to say
        let again = CadenceRules.evaluate(
            entries: run(minutes: minutes.filter { $0 <= 110 }), config: config(),
            now: base.addingTimeInterval(115 * 60), calendar: calendar, state: &state)
        XCTAssertTrue(again.isEmpty, "a stretch says something once per threshold, not per poll")

        // Past three hours: the second multiple
        let atSecond = CadenceRules.evaluate(
            entries: run(minutes: minutes), config: config(),
            now: base.addingTimeInterval(205 * 60), calendar: calendar, state: &state)
        XCTAssertEqual(atSecond.count, 1)
    }

    func testANewStretchRearmsTheCue() {
        var state = CadenceState()
        let first = Array(stride(from: 0.0, through: 100.0, by: 10.0))
        _ = CadenceRules.evaluate(entries: run(minutes: first), config: config(),
                                  now: base.addingTimeInterval(105 * 60),
                                  calendar: calendar, state: &state)
        // A real break, then a second long run
        let second = first + Array(stride(from: 200.0, through: 300.0, by: 10.0))
        let cues = CadenceRules.evaluate(entries: run(minutes: second), config: config(),
                                         now: base.addingTimeInterval(305 * 60),
                                         calendar: calendar, state: &state)
        XCTAssertEqual(cues.count, 1)
    }

    func testLateCueFiresOncePerNight() {
        var state = CadenceState()
        var cfg = config()
        cfg.stretchMinutes = 600  // keep the stretch rule out of this test
        let lateOne = base.addingTimeInterval(14 * 3600 + 600)   // 23:10
        let lateTwo = base.addingTimeInterval(15 * 3600)         // 00:00, same night
        var entries = [entry(lateOne)]
        let first = CadenceRules.evaluate(entries: entries, config: cfg,
                                          now: lateOne.addingTimeInterval(60),
                                          calendar: calendar, state: &state)
        XCTAssertEqual(first.count, 1)
        if case .late = first[0].kind {} else { XCTFail("expected a late cue") }

        entries.append(entry(lateTwo))
        let second = CadenceRules.evaluate(entries: entries, config: cfg,
                                           now: lateTwo.addingTimeInterval(60),
                                           calendar: calendar, state: &state)
        XCTAssertTrue(second.isEmpty, "the same night is one cue")
    }

    func testLateCueNeedsRecentActivityNotAMemoryOfIt() {
        var state = CadenceState()
        var cfg = config()
        cfg.stretchMinutes = 600
        let lateLastNight = base.addingTimeInterval(14 * 3600)
        let cues = CadenceRules.evaluate(entries: [entry(lateLastNight)], config: cfg,
                                         now: lateLastNight.addingTimeInterval(4 * 3600),
                                         calendar: calendar, state: &state)
        XCTAssertTrue(cues.isEmpty, "an old reading is not news about tonight")
    }

    func testStreakIsAnnouncedAtItsThresholdAndNotEveryDayAfter() {
        var cfg = config()
        cfg.stretchMinutes = 600
        var state = CadenceState()
        func entries(days: Int) -> [Entry] {
            (0..<days).map { entry(base.addingTimeInterval(-Double($0) * 86400)) }
        }
        let atSeven = CadenceRules.evaluate(entries: entries(days: 7), config: cfg, now: base,
                                            calendar: calendar, state: &state)
        XCTAssertEqual(atSeven.count, 1)
        if case .streak(let days) = atSeven[0].kind { XCTAssertEqual(days, 7) }
        else { XCTFail("expected a streak cue") }

        let atEight = CadenceRules.evaluate(entries: entries(days: 8), config: cfg, now: base,
                                            calendar: calendar, state: &state)
        XCTAssertTrue(atEight.isEmpty, "an eighth day is not a second announcement")

        let atFourteen = CadenceRules.evaluate(entries: entries(days: 14), config: cfg,
                                               now: base, calendar: calendar, state: &state)
        XCTAssertEqual(atFourteen.count, 1, "the next multiple is worth saying")
    }

    func testNoEntriesMeansNoCues() {
        var state = CadenceState()
        XCTAssertTrue(CadenceRules.evaluate(entries: [], config: config(), now: base,
                                            calendar: calendar, state: &state).isEmpty)
    }

    func testStateRoundTripsThroughItsFile() {
        let url = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("cadence-\(UUID().uuidString).json")
        defer { try? FileManager.default.removeItem(at: url) }
        var state = CadenceState()
        state.stretchID = "stretch|123"
        state.stretchFired = 2
        state.lastLateNight = "2026-08-18"
        state.streakFired = 1
        XCTAssertTrue(CadenceStore.save(state, to: url))
        XCTAssertEqual(CadenceStore.load(from: url), state)
    }
}

File: windows/tests/Redline.Core.Tests/CadenceTests.cs
// Cues about how the work is spread out. The rules pinned are the ones that keep it from
// becoming a nag: once per stretch, once per night, nothing when off or when the data is old.
using Redline.Core;

namespace Redline.Core.Tests;

public sealed class CadenceTests
{
    private static readonly TimeZoneInfo Utc = TimeZoneInfo.Utc;

    private static Entry Entry(DateTimeOffset ts) =>
        new("Claude", Guid.NewGuid().ToString(), ts, "claude-sonnet-5", 100, 10, 0, 0, 0);

    /// 2026-08-18 09:00:00 UTC
    private static readonly DateTimeOffset Base = DateTimeOffset.FromUnixTimeSeconds(1_787_043_600);

    private static List<Entry> Run(IEnumerable<double> minutes) =>
        minutes.Select(m => Entry(Base.AddSeconds(m * 60))).ToList();

    private static IEnumerable<double> Stride(double from, double through, double by)
    {
        for (var x = from; x <= through; x += by) yield return x;
    }

    // MARK: - Shapes

    [Fact]
    public void AGapLongerThanTheBreakSplitsTheStretch()
    {
        var stretches = Cadence.Stretches(Run(new double[] { 0, 5, 10, 40, 45 }));
        Assert.Equal(2, stretches.Count);
        Assert.Equal(600, stretches[0].Length);
        Assert.Equal(300, stretches[1].Length);
    }

    [Fact]
    public void AShortPauseDoesNotSplitAStretch()
    {
        // A slow turn or a build is not a break.
        Assert.Single(Cadence.Stretches(Run(new double[] { 0, 12, 24 })));
    }

    [Fact]
    public void StretchLengthIsMeasuredToTheLastActivityNotToNow()
    {
        var stretch = Cadence.Stretches(Run(new double[] { 0, 10, 20, 30 })).LastOrDefault();
        // a counter that climbs while nothing happens is not a measurement
        Assert.Equal(1800, stretch?.Length);
    }

    [Fact]
    public void CurrentStretchClosesOnceTheGapHasPassed()
    {
        var entries = Run(new double[] { 0, 10, 20, 30 });
        Assert.NotNull(Cadence.Current(entries, now: Base.AddSeconds(35 * 60)));
        Assert.Null(Cadence.Current(entries, now: Base.AddSeconds(60 * 60)));
    }

    [Fact]
    public void StreakCountsBackFromTodayAndStopsAtTheFirstGap()
    {
        var entries = new[] { 0, 1, 2, 4, 5 }.Select(d => Entry(Base.AddSeconds(-d * 86400.0))).ToList();
        Assert.Equal(3, Cadence.Streak(entries, Base, Utc));
    }

    [Fact]
    public void StreakIsZeroWhenTodayHasNothing()
    {
        var entries = new[] { Entry(Base.AddSeconds(-2 * 86400)) };
        Assert.Equal(0, Cadence.Streak(entries, Base, Utc));
    }

    [Fact]
    public void HourHistogramBucketsByLocalHour()
    {
        var hours = Cadence.ByHourOfDay(Run(new double[] { 0, 1, 120 }), Utc);
        Assert.Equal(220, hours[9]);
        Assert.Equal(110, hours[11]);
        Assert.Equal(330, hours.Sum());
    }

    [Fact]
    public void NightRollsAtFiveNotAtMidnight()
    {
        // 01:30 belongs to the evening that ran into it.
        var lateEvening = Base.AddSeconds(14 * 3600);       // 23:00
        var afterMidnight = Base.AddSeconds(16.5 * 3600);   // 01:30 next day
        Assert.Equal(Cadence.NightKey(lateEvening, Utc), Cadence.NightKey(afterMidnight, Utc));
    }

    /// Windows port only: the injected zone decides the local hour, not the machine's own.
    [Fact]
    public void LocalHourFollowsTheInjectedTimeZone()
    {
        var tokyo = TimeZoneInfo.CreateCustomTimeZone("UTC+9", TimeSpan.FromHours(9), "UTC+9", "UTC+9");
        var hours = Cadence.ByHourOfDay(Run(new double[] { 0 }), tokyo);
        Assert.Equal(110, hours[18]);
        // 16:00 UTC on the 18th is 01:00 on the 19th in Tokyo: a late night, not a new day
        var late = Base.AddSeconds(7 * 3600);
        Assert.Equal(Cadence.NightKey(Base.AddSeconds(-4 * 3600), tokyo), Cadence.NightKey(late, tokyo));
    }

    // MARK: - Cues

    private static Config Cfg() => new()
    {
        MindfulCues = true,
        StretchMinutes = 90,
        LateHour = 23,
        StreakDays = 7,
    };

    [Fact]
    public void CuesAreSilentWhenTheSettingIsOff()
    {
        var state = new CadenceState();
        var off = Cfg();
        off.MindfulCues = false;
        var entries = Run(Stride(0, 240, 10));
        var cues = CadenceRules.Evaluate(entries, off, state, Base.AddSeconds(240 * 60), Utc);
        Assert.Empty(cues);
    }

    [Fact]
    public void AStretchIsAnnouncedOnceAndThenAtTheNextMultiple()
    {
        var state = new CadenceState();
        var minutes = Stride(0, 200, 10).ToList();
        // At 100 minutes: one threshold reached
        var atFirst = CadenceRules.Evaluate(Run(minutes.Where(m => m <= 100)), Cfg(), state,
            Base.AddSeconds(105 * 60), Utc);
        Assert.Single(atFirst);
        Assert.IsType<CadenceCueKind.Stretch>(atFirst[0].Kind);

        // Ten minutes later, still the same run: nothing new to say
        var again = CadenceRules.Evaluate(Run(minutes.Where(m => m <= 110)), Cfg(), state,
            Base.AddSeconds(115 * 60), Utc);
        Assert.Empty(again); // a stretch says something once per threshold, not per poll

        // Past three hours: the second multiple
        var atSecond = CadenceRules.Evaluate(Run(minutes), Cfg(), state, Base.AddSeconds(205 * 60), Utc);
        Assert.Single(atSecond);
    }

    [Fact]
    public void ANewStretchRearmsTheCue()
    {
        var state = new CadenceState();
        var first = Stride(0, 100, 10).ToList();
        CadenceRules.Evaluate(Run(first), Cfg(), state, Base.AddSeconds(105 * 60), Utc);
        // A real break, then a second long run
        var second = first.Concat(Stride(200, 300, 10));
        var cues = CadenceRules.Evaluate(Run(second), Cfg(), state, Base.AddSeconds(305 * 60), Utc);
        Assert.Single(cues);
    }

    [Fact]
    public void LateCueFiresOncePerNight()
    {
        var state = new CadenceState();
        var cfg = Cfg();
        cfg.StretchMinutes = 600; // keep the stretch rule out of this test
        var lateOne = Base.AddSeconds(14 * 3600 + 600);  // 23:10
        var lateTwo = Base.AddSeconds(15 * 3600);        // 00:00, same night
        var entries = new List<Entry> { Entry(lateOne) };
        var first = CadenceRules.Evaluate(entries, cfg, state, lateOne.AddSeconds(60), Utc);
        Assert.Single(first);
        Assert.IsType<CadenceCueKind.Late>(first[0].Kind);

        entries.Add(Entry(lateTwo));
        var second = CadenceRules.Evaluate(entries, cfg, state, lateTwo.AddSeconds(60), Utc);
        Assert.Empty(second); // the same night is one cue
    }

    [Fact]
    public void LateCueNeedsRecentActivityNotAMemoryOfIt()
    {
        var state = new CadenceState();
        var cfg = Cfg();
        cfg.StretchMinutes = 600;
        var lateLastNight = Base.AddSeconds(14 * 3600);
        var cues = CadenceRules.Evaluate(new[] { Entry(lateLastNight) }, cfg, state,
            lateLastNight.AddSeconds(4 * 3600), Utc);
        Assert.Empty(cues); // an old reading is not news about tonight
    }

    [Fact]
    public void StreakIsAnnouncedAtItsThresholdAndNotEveryDayAfter()
    {
        var cfg = Cfg();
        cfg.StretchMinutes = 600;
        var state = new CadenceState();
        List<Entry> Entries(int days) =>
            Enumerable.Range(0, days).Select(d => Entry(Base.AddSeconds(-d * 86400.0))).ToList();

        var atSeven = CadenceRules.Evaluate(Entries(7), cfg, state, Base, Utc);
        Assert.Single(atSeven);
        var streak = Assert.IsType<CadenceCueKind.Streak>(atSeven[0].Kind);
        Assert.Equal(7, streak.Days);

        var atEight = CadenceRules.Evaluate(Entries(8), cfg, state, Base, Utc);
        Assert.Empty(atEight); // an eighth day is not a second announcement

        var atFourteen = CadenceRules.Evaluate(Entries(14), cfg, state, Base, Utc);
        Assert.Single(atFourteen); // the next multiple is worth saying
    }

    [Fact]
    public void NoEntriesMeansNoCues()
    {
        var state = new CadenceState();
        Assert.Empty(CadenceRules.Evaluate(new List<Entry>(), Cfg(), state, Base, Utc));
    }

    [Fact]
    public void StateRoundTripsThroughItsFile()
    {
        var path = Path.Combine(Path.GetTempPath(), $"cadence-{Guid.NewGuid()}.json");
        try
        {
            var state = new CadenceState
            {
                StretchID = "stretch|123",
                StretchFired = 2,
                LastLateNight = "2026-08-18",
                StreakFired = 1,
            };
            Assert.True(CadenceStore.Save(state, path));
            Assert.Equal(state, CadenceStore.Load(path));
        }
        finally { try { File.Delete(path); } catch { } }
    }
}
